src/bin/mir-json-translate-libs.rs

```diff
@@ -0,0 +1,701 @@
+}
+
+/// Module to hide field of newtype [CrateName].
+mod crate_name {
+    use std::fmt;
+
+    use serde::Deserialize;
+
+    /// Newtype wrapper around [String] which enforces that all `-` are replaced
+    /// with `_`. Package names often contain `-`, but it is not a valid
+    /// character in Rust identifiers, and crate names must be valid Rust
+    /// identifiers since they are automatically exposed as module names in Rust
+    /// code. Therefore cargo converts `-` to `_` when deriving crate names from
+    /// package names.
+    ///
+    /// Construct with [From<&str>].
+    #[derive(Clone, Deserialize, PartialEq)]
+    pub struct CrateName(String);
+
+    impl From<&str> for CrateName {
+        fn from(name: &str) -> Self {
+            Self(name.replace("-", "_"))
+        }
+    }
+
+    impl From<CrateName> for String {
+        fn from(crate_name: CrateName) -> String {
+            crate_name.0
+        }
+    }
+
+    impl fmt::Display for CrateName {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            self.0.fmt(f)
+        }
+    }
+}
+use crate_name::CrateName;
+
+/// Our version of the unit graph, which allows for adding our custom libraries
+/// and easily converting to mir-json commands.
+struct CustomUnitGraph {
+    units: Vec<CustomUnit>,
+    roots: Vec<usize>,
+}
+
+struct CustomUnit {
+    target: CustomTarget,
+    dependencies: Vec<UnitGraphDependency>,
+}
+
+enum CustomTarget {
+    /// A library to be compiled for the target architecture. We really only
+    /// care about the target information for these, so only this variant has
+    /// fields.
+    TargetLib(CustomTargetLib),
+    /// A dependency to be compiled for the host, like a build script or proc
+    /// macro.
+    HostDep,
+    /// A binary that is not a [CustomTarget::HostDep].
+    FinalBin,
+}
+
+/// The fields here are a combination of the fields we care about from [Target],
+/// [cargo_metadata::Artifact], and [cargo_metadata::BuildScript]. Each
+/// corresponds to some input to `rustc`.
+#[derive(Clone)]
+struct CustomTargetLib {
+    /// `--crate-name`
+    crate_name: CrateName,
+    /// source file argument
+    src_path: Utf8PathBuf,
+    /// `--edition`
+    edition: Edition,
+    /// `-l`
+    linked_libs: Vec<Utf8PathBuf>,
+    /// `-L`
+    linked_paths: Vec<Utf8PathBuf>,
+    /// `--cfg`
+    cfgs: Vec<String>,
+    /// environment variables
+    env: Vec<(String, String)>,
+}
+
+impl CustomTarget {
+    fn is_target_lib_with_crate_name(&self, crate_name: &CrateName) -> bool {
+        match self {
+            Self::TargetLib(lib) if lib.crate_name == *crate_name => true,
+            _ => false,
+        }
+    }
+}
+
+/// A library that will be compiled by mir-json.
+struct MirJsonLib {
+    target: CustomTargetLib,
+    /// `(x, y)` is passed as `--extern x=liby.rlib`. Usually `x` is the same
+    /// as `y` but not always.
+    dependencies: Vec<(CrateName, CrateName)>,
```

Comment:
Is the `y` part actually required to be a `CrateName`?  IIRC Cargo typically uses `-` in dependency filenames, like `libfoo-00112233.rlib`

Reply:
Ah, good point